### PR TITLE
[FLINK-18641][checkpointing] Fix CheckpointCoordinator to work with ExternallyInducedSource

### PR DESCRIPTION
**Pull request by becketqin**
## What is the purpose of the change
This patch fixes the `CheckpointCoordinator` to make it work with `ExternallyInducedSource`  in cases that the task snapshots are triggered by the external systems via master hooks, rather than the checkpoint coordinator.

The problem in the current code is that when the task snapshots are triggered externally via the master hooks, the checkpoint coordinator may receive all the acks from the tasks before the master state snapshot completes. And this leads to checkpoint failure. The fix is to only finalize the checkpoint when all of the operator coordinator checkpoint, master snapshots and task snapshots are fully taken.

This patch also fixes the order of component checkpoint by putting the `OperatorCoordinator` checkpoint to before the master hooks invocation.

## Brief change log
- Finalize the checkpoint when the checkpoint is fully acknowledged, which can be triggered from either the acknowledge of task snapshot or master snapshot.
- Adjust the checkpoint order from `(master hooks, OperatorCoordinators) -> tasks` to `OperatorCoordinators -> master hooks -> tasks`.

## Verifying this change
This change added tests and can be verified by running `CheckpointCoordinatorTest#testTaskCheckpointTriggeredByMasterHooks()`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java [hunk 1]
```diff
@@ -576,7 +583,22 @@ private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
 										request.advanceToEndOfTime);
 
 									coordinatorsToCheckpoint.forEach((ctx) -> ctx.afterSourceBarrierInjection(checkpointId));
-
+									// It is possible that the tasks has finished checkpointing at this point.
+									// So we need to complete this pending checkpoint.
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:587]: In an essence, this is a race condition between sending out snapshot state calls, and receiving ack's. And the race condition is caused by the fact, that `CheckpointCoordinator#receiveAcknowledgeMessage` is executed in the `ioExecutor`, while this logic here is happening on the timer thread?

If so, can we write this as a comment here?

Secondly, @ifndef-SleePy, would this problem go away with the last step of your refactor? I guess the idea should be, that both of those things would be executed in job manager's main thread executor, right?

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java [hunk 2]
```diff
@@ -2260,6 +2263,121 @@ public void testSavepointScheduledInUnalignedMode() throws Exception {
...
+			CompletedCheckpoint success = coord.getSuccessfulCheckpoints().get(0);
+			assertEquals(jid, success.getJobId());
+			assertEquals(2, success.getOperatorStates().size());
+
+			coord.shutdown(JobStatus.FINISHED);
+		}
+		catch (Throwable e) {
+			e.printStackTrace();
+			fail(e.getMessage());
+		}
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java:2378]: This way of handling and catching of an exception is a left over of some workaround of an old bug (JUnit? Travis? I have no idea). Nowadays there is no need for this, so you could remove this try/catch block - it would also improve error message, as you wouldn't be hiding the original stack trace.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java [hunk 3]
```diff
@@ -2260,6 +2263,121 @@ public void testSavepointScheduledInUnalignedMode() throws Exception {
 		}
 	}
 
+	/**
+	 * Test that the checkpoint still behave correctly when the task checkpoint is triggered by the
+	 * master hooks and finished before the master checkpoint.
+	 */
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java:2269]: So this test is covering for both bug fixes?

Comment (becketqin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java:2269]: The test covers the checkpoint failure issue. The other case is sort of indirectly tested by other tests which ensures all the `OperatorCoordinators` have finished snapshots before moving on to the next step. We can add another test to explicitly test the ordering issue.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java [hunk 4]
```diff
@@ -530,15 +530,22 @@ private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
...
 									coordinatorsToCheckpoint, pendingCheckpoint, timer),
 							timer);
 
+			// We have to take the snapshot of the master hooks after the coordinator checkpoints has completed.
+			// This is to ensure the tasks are checkpointed after the OperatorCoordinators in case
+			// ExternallyInducedSource is used.
+			final CompletableFuture<?> masterStatesComplete = coordinatorCheckpointsComplete
+					.thenComposeAsync(ignored -> {
+						PendingCheckpoint checkpoint =
+							FutureUtils.getWithoutException(pendingCheckpointCompletableFuture);
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:545]: Ok, I see how it's supposed to be working. But in that case there is a missing `checkState` here.

```
PendingCheckpoint checkpoint =
	FutureUtils.getWithoutException(pendingCheckpointCompletableFuture);

Preconditions.checkState(
	checkpoint != null || throwable != null,
	"?????????.");
```
% maybe code deduplication with the same get and check from  L553:L558. It's better to fail early, with `IllegalStateException` compared to `NPE` in `snapshotMasterState`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java (hunk 1 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:587]: Ok, but at least `synchronized (lock) {` wouldn't be needed anymore, right?

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java (hunk 3 above)
Comment (pnowojski) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java:2269]: > The other case is sort of indirectly tested by other tests which ensures all the OperatorCoordinators have finished snapshots before moving on to the next step. 

But as you haven't modified those tests, and the master build was green, it seems like those tests were passing with the incorrect code.

Doesn't it mean we are still missing test coverage?

If so, that would mean splitting this change into to two separate bug fixes, with two separate test fixes would be much easier to understand and follow, which unit test is covering for what bug.

Comment (pnowojski) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java:2269]: > I end up with testing in one case for simplicity

It looks like the first bug fix from
```
final CompletableFuture<?> masterStatesComplete = pendingCheckpointCompletableFuture	
					.thenCompose(this::snapshotMasterState);
```
to
```
			final CompletableFuture<?> masterStatesComplete = coordinatorCheckpointsComplete
					.thenComposeAsync(ignored -> {
						PendingCheckpoint checkpoint =
							FutureUtils.getWithoutException(pendingCheckpointCompletableFuture);
						return snapshotMasterState(checkpoint);
					}, timer);
```
doesn't have a test coverage.

Could you split the two bug fixes into two commits? Besides that having both fixes in one commit is confusing it's also pretending that both are covered via the same unit test.

Comment (becketqin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java:2269]: I haven't got time to update the PR yet. The upcoming update has separated the PRs into two commits and added test coverage for both cases.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java [hunk 5]
```diff
@@ -2260,6 +2263,121 @@ public void testSavepointScheduledInUnalignedMode() throws Exception {
...
+			AtomicReference<Long> checkpointIdRef = new AtomicReference<>();
+
+			OperatorID opID1 = OperatorID.fromJobVertexID(vertex1.getJobvertexId());
+			OperatorID opID2 = OperatorID.fromJobVertexID(vertex2.getJobvertexId());
+			TaskStateSnapshot taskOperatorSubtaskStates1 = mock(TaskStateSnapshot.class);
+			TaskStateSnapshot taskOperatorSubtaskStates2 = mock(TaskStateSnapshot.class);
+			OperatorSubtaskState subtaskState1 = mock(OperatorSubtaskState.class);
+			OperatorSubtaskState subtaskState2 = mock(OperatorSubtaskState.class);
+			when(taskOperatorSubtaskStates1.getSubtaskStateByOperatorID(opID1)).thenReturn(subtaskState1);
+			when(taskOperatorSubtaskStates2.getSubtaskStateByOperatorID(opID2)).thenReturn(subtaskState2);
```
[the remark this answers is not in the document]
Comment (becketqin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java:2294]: This is a legacy class, so I just followed the pattern to make the tests consistent in this class. Personally I prefer removing mockito in a separate ticket at the unit test class granularity instead of at per test method granularity.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java [hunk 6]
```diff
@@ -2260,6 +2263,121 @@ public void testSavepointScheduledInUnalignedMode() throws Exception {
...
 	}
 
+	/**
+	 * Test that the checkpoint still behave correctly when the task checkpoint is triggered by the
+	 * master hooks and finished before the master checkpoint.
+	 */
+	@Test
+	public void testTaskCheckpointTriggeredByMasterHooks() {
+		try {
+			final JobID jid = new JobID();
```
[the remark this answers is not in the document]
Comment (becketqin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java:2273]: I don't have a strong opinion on this. `jid` and `coord` are common names used by other test methods in this class, and I also feel they are intuitive enough, so I just followed the same convention.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java [hunk 7]
```diff
@@ -2260,6 +2263,121 @@ public void testSavepointScheduledInUnalignedMode() throws Exception {
...
+			});
+
+			assertEquals(0, coord.getNumberOfPendingCheckpoints());
+			assertEquals(0, coord.getNumberOfRetainedSuccessfulCheckpoints());
+			assertEquals(0, manuallyTriggeredScheduledExecutor.getScheduledTasks().size());
+
+			// trigger the first checkpoint. this should succeed
+			final CompletableFuture<CompletedCheckpoint> checkpointFuture = coord.triggerCheckpoint(false);
+			manuallyTriggeredScheduledExecutor.triggerAll();
+			assertFalse(checkpointFuture.isCompletedExceptionally());
```
[the remark this answers is not in the document]
Comment (becketqin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java:2349]: Technically speaking, `future.get()` does not provide the same check here. It is also possible that the future is not completed after `triggerAll()` is invoked, either expected or not. In this case, `future.get()` will block forever.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:545]: > the behavior is guaranteed by CompletableFuture, the assertion here would essentially be verifying CompletableFuture,

Not exactly. It would be verifying that you chained a couple of futures and callbacks correctly. That the callback `foo()` is using `future1` result and is triggered once `future2` completes, and that `future1` and `future2` are chained (or am I still mis understanding this code?). Java library doesn't guarantee you that, but your code that is chaining the futures does. Which is outside of the `foo()`'s control, so from `foo()`s perspective, that's an external assumption, and falls under:

> ensure the interface contract with users are not broken.

Where "users" are function's callers.

And as I wrote before. If something violates this assumption, and even if some unit test fail, it's a bit easier to understand a `checkState` compared to `NPE`. Note performance overhead of one if check doesn't matter here at all. Also it's harder of `checkState` to become outdated and misleading over time.

If you have so strong feelings about, put a comment, but I do not see any drawback of replacing a comment with a `checkState` with the same comment but as a message here.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java:2273]: It's not a big issue, but the "convention" dates to 2015 and we are trying to change it recently. They are intuitive if you are reading or writing whole test, but not necessarily when reading a call like
```
getCheckpointCoordinator(jid, vertex1, vertex2);
```
newer code tries to get rid of the abbreviations here.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java:2294]: If you prefer, removing the mockito in all of those unit tests would be really great and appreciated. But if we don't have time to do it, please do not add instant technological debt :( Like in many other places, we are trying to incrementally improve code quality and this goes in the opposite direction.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java:2349]: Maybe add a helper method in `FutureUtils#throwIfCompletedExceptionally(checkpointFuture)`?  

(I've found this when I was trying out the test coverage of this unit test, and `assertFalse` is not very helpful here)

**Comment by becketqin** [comment]
@pnowojski Thanks for the review. I updated the patch. Some clarifications:

1. Regarding the `checkState()` calls, I still feel a little weird to have that just to verify our own code is correct. So I just put a comment there.
2. I removed the mockito usages in `CheckpointCoordinatorTest` class and replaced them with real classes. These classes somewhat look like our own version of mockito, though. So I am not sure if the `no-mock` rule is something beneficial in all cases. Or should we still leverage that when it helps reduce the verbosity.
3. The abbreviated variable names are replaced with camel format. Personally I feel the abbreviated names are less verbose and thus more readable, but I don't have a strong opinion on that.

BTW, since this ticket is a blocker for 1.11.2 release, please feel free to ping me if you have further comments. Thanks.

**Comment by becketqin** [comment]
@flinkbot run azure

**Comment by becketqin** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java [hunk 8]
```diff
@@ -955,7 +978,7 @@ public boolean receiveAcknowledgeMessage(AcknowledgeCheckpoint message, String t
 						LOG.debug("Received acknowledge message for checkpoint {} from task {} of job {} at {}.",
 							checkpointId, message.getTaskExecutionId(), message.getJob(), taskManagerLocationInfo);
 
-						if (checkpoint.areTasksFullyAcknowledged()) {
+						if (checkpoint.isFullyAcknowledged()) {
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:991]: Is this change part of the same bug fix as the changes above?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java [hunk 9]
```diff
@@ -528,15 +528,25 @@ private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
...
-
 			final CompletableFuture<?> coordinatorCheckpointsComplete = pendingCheckpointCompletableFuture
 					.thenComposeAsync((pendingCheckpoint) ->
 							OperatorCoordinatorCheckpoints.triggerAndAcknowledgeAllCoordinatorCheckpointsWithCompletion(
 									coordinatorsToCheckpoint, pendingCheckpoint, timer),
 							timer);
 
+			// We have to take the snapshot of the master hooks after the coordinator checkpoints has completed.
+			// This is to ensure the tasks are checkpointed after the OperatorCoordinators in case
+			// ExternallyInducedSource is used.
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:539]: nit: could you copy paste this comment to the commits description?

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java [hunk 10]
```diff
@@ -2436,6 +2442,126 @@ public Integer deserialize(int version, byte[] serialized) throws IOException {
 		coord.shutdown(JobStatus.FINISHED);
 	}
 
+	@Test
+	public void testCompleteCheckpointFailureWithExternallyInducedSource() throws Exception {
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java:2447]: why is this added in a separate commit?

**Review by pnowojski — COMMENTED** [review]
Thanks for the update @becketqin. I've left a couple of comments/questions.

**Review by pnowojski — COMMENTED** [review]
As discussed online, the final code LGTM. The commits structure needs a bit of adjustment and I'm going to fix this manually before merging.

**Comment by pnowojski** [comment]
Merged, thanks for the fix @becketqin :)

**Comment by zhuzhurk** [comment]
Thanks for preparing and merging this fix! @becketqin @pnowojski
